File: moovin_agents_SDK/SilverAI_ari.py


"""
SilverAI_ari.py
--------------
App ARI minimalista para:
- Responder llamadas que entren a la app 'silverai' 
- Crear un bridge mixing/proxy_media
- Crear un canal ExternalMedia hacia 127.0.0.1:40000
- Añadir ambos canales al bridge
- Limpiar recursos al colgar

Es escuchado por el SDK
"""

import os
import sys
import time
import signal
import logging
from typing import Dict, Optional
import ari 



# --------------------------
# Config vía variables de entorno
# --------------------------
ARI_URL            = os.getenv("ARI_URL")
ARI_USER           = os.getenv("ARI_USER", "asterisk")
ARI_PASS           = os.getenv("ARI_PASS", "asterisk")
ARI_APP            = os.getenv("ARI_APP", "app") 

EXTERNAL_HOST         = os.getenv("EXTERNAL_HOST")  # host:port en gateway SDK
EXTERNAL_FORMAT       = os.getenv("EXTERNAL_FORMAT", "alaw")           # alaw | ulaw | slin16
EXTERNAL_TRANSPORT    = os.getenv("EXTERNAL_TRANSPORT", "udp")         # udp | tcp | websocket
EXTERNAL_ENCAPSULATION= os.getenv("EXTERNAL_ENCAPSULATION", "rtp")     # rtp | audiosocket | none
EXTERNAL_DIRECTION    = os.getenv("EXTERNAL_DIRECTION", "both")        # both | in | out

MAX_CALL_MS        = int(os.getenv("MAX_CALL_MS", "0"))
LOG_LEVEL = os.getenv("LOG_LEVEL", "INFO").upper()

# --------------------------
# Logging
# --------------------------
logging.basicConfig(
    level=LOG_LEVEL,
    format="%(asctime)s | %(levelname)s | %(message)s",
)
log = logging.getLogger("silverai_ari")

# --------------------------
# Estado en memoria
# --------------------------
class CallState:
    """Estructura para trackear recursos por canal SIP."""
    def __init__(self, sip_channel_id: str):
        self.sip_channel_id = sip_channel_id
        self.bridge_id: Optional[str] = None
        self.ext_channel_id: Optional[str] = None
        self.started_at_ms = int(time.time() * 1000)

CALLS: Dict[str, CallState] = {}
EXT_TO_SIP: Dict[str, str] = {}

client: Optional[ari.Client] = None
_running = True

# --------------------------
# Utilidades
# --------------------------
def ms_since(ts_ms: int) -> int:
    return int(time.time() * 1000) - ts_ms

def safe_get(dct: dict, *path, default=None):
    cur = dct
    for p in path:
        if not isinstance(cur, dict) or p not in cur:
            return default
        cur = cur[p]
    return cur

# --------------------------
# Manejadores de eventos ARI
# --------------------------
def on_stasis_start(channel_obj, event):
    """
    Se dispara cuando un canal entra a la app ARI (StasisStart).
    Puede ser:
      - El canal SIP entrante 
      - El canal ExternalMedia
    """
    ch_id = safe_get(event, "channel", "id")
    ch_name = safe_get(event, "channel", "name")
    log.info(f"StasisStart: channel_id={ch_id} name={ch_name}")

    is_external = ch_name and (ch_name.startswith("UnicastRTP") or ch_name.startswith("AudioSocket") or ch_name.startswith("WebSocketChannel"))
    if is_external:
        sip_id = EXT_TO_SIP.get(ch_id)
        if not sip_id:
            log.info(f"External channel {ch_id} llegó antes de mapear; pequeño delay para resolver bridge…")
            time.sleep(0.3)
            sip_id = EXT_TO_SIP.get(ch_id)

        if not sip_id or sip_id not in CALLS:
            log.warning(f"No encuentro SIP asociado para external channel {ch_id}. Ignoro por ahora.")
            return

        state = CALLS[sip_id]
        if not state.bridge_id:
            log.warning(f"No hay bridge todavía para SIP {sip_id}. Ignoro external {ch_id}.")
            return

        try:
            bridge = client.bridges.get(bridgeId=state.bridge_id)
            bridge.addChannel(channel=ch_id)
            log.info(f"Añadido external channel {ch_id} al bridge {state.bridge_id}")
        except Exception as e:
            log.error(f"Error añadiendo external {ch_id} al bridge {state.bridge_id}: {e}")
        return
    sip_id = ch_id
    CALLS[sip_id] = CallState(sip_channel_id=sip_id)

    channel = client.channels.get(channelId=sip_id)
    try:
        channel.answer()
        log.info(f"SIP {sip_id} contestado")
        bridge = client.bridges.create(type="mixing,proxy_media")
        CALLS[sip_id].bridge_id = bridge.id
        log.info(f"Bridge creado: {bridge.id}")
        bridge.addChannel(channel=sip_id)
        log.info(f"SIP {sip_id} agregado al bridge {bridge.id}")

        # Cortar al límite de duración (Desactivado)
        if MAX_CALL_MS > 0:
            def _enforce_limit():
                try:
                    if sip_id in CALLS and ms_since(CALLS[sip_id].started_at_ms) > MAX_CALL_MS:
                        log.info(f"Max call time superado para {sip_id}. Colgando.")
                        client.channels.hangup(channelId=sip_id)
                except Exception as e:
                    log.warning(f"Al forzar límite: {e}")
            client.add_periodic_task(3.0, _enforce_limit)

        # Canar ExternalMedia apuntado al Getaway
        params = {
            "app": ARI_APP,
            "external_host": EXTERNAL_HOST,      
            "format": EXTERNAL_FORMAT,           
            "transport": EXTERNAL_TRANSPORT,      
            "encapsulation": EXTERNAL_ENCAPSULATION, 
            "connection_type": "client",         
            "direction": EXTERNAL_DIRECTION    
        }
        ext_ch = client.channels.externalMedia(**params)
        CALLS[sip_id].ext_channel_id = ext_ch.id
        EXT_TO_SIP[ext_ch.id] = sip_id
        log.info(f"ExternalMedia creado: {ext_ch.id} -> {EXTERNAL_HOST} ({EXTERNAL_FORMAT}/{EXTERNAL_TRANSPORT}/{EXTERNAL_ENCAPSULATION})")

    except Exception as e:
        log.error(f"Error preparando llamada para SIP {sip_id}: {e}")
        try:
            channel.hangup()
        except Exception:
            pass
        cleanup_call(sip_id)

def on_stasis_end(channel_obj, event):
    """ Limpieza cuando un canal sale de la app (colgado). """
    ch_id = safe_get(event, "channel", "id")
    ch_name = safe_get(event, "channel", "name")
    log.info(f"StasisEnd: channel_id={ch_id} name={ch_name}")

    if ch_id in EXT_TO_SIP:
        sip_id = EXT_TO_SIP.pop(ch_id, None)
        log.info(f"External {ch_id} terminó (SIP asociado: {sip_id})")
        return

    cleanup_call(ch_id)

def cleanup_call(sip_id: str):
    """ Cierra bridge, cuelga external (si existe) y borra estado. """
    state = CALLS.pop(sip_id, None)
    if not state:
        return

    if state.ext_channel_id:
        try:
            client.channels.hangup(channelId=state.ext_channel_id)
            log.info(f"External {state.ext_channel_id} colgado")
        except Exception:
            pass
        EXT_TO_SIP.pop(state.ext_channel_id, None)

    if state.bridge_id:
        try:
            br = client.bridges.get(bridgeId=state.bridge_id)
            br.destroy()
            log.info(f"Bridge {state.bridge_id} destruido")
        except Exception:
            pass

    try:
        client.channels.hangup(channelId=sip_id)
    except Exception:
        pass

# --------------------------
# Señales del sistema
# --------------------------
def _signal_handler(sig, frame):
    global _running
    log.info(f"Señal {sig} recibida. Cerrando…")
    _running = False
    try:
        for sip_id in list(CALLS.keys()):
            cleanup_call(sip_id)
    finally:
        try:
            client.close()
        except Exception:
            pass
        sys.exit(0)

# --------------------------
# Main
# --------------------------
def main():
    global client
    try:
        log.info(f"Conectando a ARI: {ARI_URL} (app={ARI_APP})")
        client = ari.connect(ARI_URL, ARI_USER, ARI_PASS)
    except Exception as e:
        log.error(f"No se pudo conectar a ARI: {e}")
        sys.exit(1)

    client.on_channel_event('StasisStart', on_stasis_start)
    client.on_channel_event('StasisEnd', on_stasis_end)

    signal.signal(signal.SIGINT, _signal_handler)
    signal.signal(signal.SIGTERM, _signal_handler)

    log.info("Escuchando eventos…")
    try:
        client.run(apps=[ARI_APP])
    except KeyboardInterrupt:
        _signal_handler("SIGINT", None)
    except Exception as e:
        log.error(f"Fallo run(): {e}")
        _signal_handler("SIGTERM", None)

if __name__ == "__main__":
    main()

File: moovin_agents_SDK/SilverAI_ari.js
// App ARI minimalista para:
// - Responder llamadas que entren a la app ARI (env ARI_APP)
// - Crear un bridge mixing/proxy_media
// - Crear un canal ExternalMedia hacia EXTERNAL_HOST
// - Añadir ambos canales al bridge
// - Limpiar recursos al colgar
//
// --------------------------------------------

require('dotenv').config();
const ari = require('ari-client');

const {
  ARI_URL,
  ARI_USER = 'asterisk',
  ARI_PASS = 'asterisk',
  ARI_APP = 'app',

  EXTERNAL_HOST,
  EXTERNAL_FORMAT = 'alaw',         
  EXTERNAL_TRANSPORT = 'udp',        
  EXTERNAL_ENCAPSULATION = 'rtp',    
  EXTERNAL_DIRECTION = 'both',       

  MAX_CALL_MS = '0',
  LOG_LEVEL = 'INFO',
} = process.env;

const LEVELS = ['ERROR', 'WARN', 'INFO', 'DEBUG'];
const CUR_LEVEL_IDX = Math.max(0, LEVELS.indexOf(String(LOG_LEVEL).toUpperCase()));
const log = {
  error: (msg) => CUR_LEVEL_IDX >= 0 && console.error(ts(), '| ERROR |', msg),
  warn:  (msg) => CUR_LEVEL_IDX >= 1 && console.warn (ts(), '| WARN  |', msg),
  info:  (msg) => CUR_LEVEL_IDX >= 2 && console.log  (ts(), '| INFO  |', msg),
  debug: (msg) => CUR_LEVEL_IDX >= 3 && console.log  (ts(), '| DEBUG |', msg),
};
function ts(){ return new Date().toISOString(); }

// --- Estado en memoria ---
class CallState {
  constructor(sipChannelId) {
    this.sipChannelId = sipChannelId;
    this.bridgeId = null;
    this.extChannelId = null;
    this.startedAtMs = Date.now();
    this.limitTimer = null; // para MAX_CALL_MS
  }
}

const CALLS = new Map();        // sipId -> CallState
const EXT_TO_SIP = new Map();   // extChannelId -> sipId
let client;                     // ari-client instance

// --- Utilidades ---
function msSince(tsMs) {
  return Date.now() - tsMs;
}

function safeGet(obj, pathArr, def = undefined) {
  let cur = obj;
  for (const p of pathArr) {
    if (!cur || typeof cur !== 'object' || !(p in cur)) return def;
    cur = cur[p];
  }
  return cur;
}

function isExternalMediaName(name) {
  if (!name) return false;
  return name.startsWith('UnicastRTP') || name.startsWith('AudioSocket') || name.startsWith('WebSocketChannel');
}

// --- Handlers ARI ---
async function onStasisStart(event, channel) {
  const chId   = safeGet(event, ['channel', 'id']);
  const chName = safeGet(event, ['channel', 'name']);
  log.info(`StasisStart: channel_id=${chId} name=${chName}`);

  // Caso: canal ExternalMedia (llega a la app también)
  if (isExternalMediaName(chName)) {
    let sipId = EXT_TO_SIP.get(chId);
    if (!sipId) {
      log.info(`External ${chId} llegó antes de mapear; esperando 300ms…`);
      await delay(300);
      sipId = EXT_TO_SIP.get(chId);
    }

    if (!sipId || !CALLS.has(sipId)) {
      log.warn(`No encuentro SIP asociado para external ${chId}. Ignoro por ahora.`);
      return;
    }

    const state = CALLS.get(sipId);
    if (!state.bridgeId) {
      log.warn(`No hay bridge todavía para SIP ${sipId}. Ignoro external ${chId}.`);
      return;
    }

    try {
      const bridge = await client.bridges.get({ bridgeId: state.bridgeId });
      await bridge.addChannel({ channel: chId });
      log.info(`Añadido external ${chId} al bridge ${state.bridgeId}`);
    } catch (e) {
      log.error(`Error añadiendo external ${chId} al bridge ${state.bridgeId}: ${e.message}`);
    }
    return;
  }

  // Caso: canal SIP entrante
  const sipId = chId;
  CALLS.set(sipId, new CallState(sipId));
  const ch = await client.channels.get({ channelId: sipId });

  try {
    // Contestar
    await ch.answer();
    log.info(`SIP ${sipId} contestado`);

    // Crear bridge mixing + proxy_media
    const bridge = await client.bridges.create({ type: 'mixing,proxy_media' });
    CALLS.get(sipId).bridgeId = bridge.id;
    log.info(`Bridge creado: ${bridge.id}`);

    await bridge.addChannel({ channel: sipId });
    log.info(`SIP ${sipId} agregado al bridge ${bridge.id}`);

    // Límite de duración
    const maxMs = parseInt(MAX_CALL_MS, 10) || 0;
    if (maxMs > 0) {
      const st = CALLS.get(sipId);
      st.limitTimer = setInterval(async () => {
        try {
          if (!CALLS.has(sipId)) return;
          if (msSince(st.startedAtMs) > maxMs) {
            log.info(`Max call time superado para ${sipId}. Colgando.`);
            await client.channels.hangup({ channelId: sipId });
          }
        } catch (err) {
          log.warn(`Al forzar límite: ${err.message}`);
        }
      }, 3000);
    }

    // Crear ExternalMedia apuntando a gateway
    const params = {
      app: ARI_APP,
      external_host: EXTERNAL_HOST,       // host:port en el gateway/SDK
      format: EXTERNAL_FORMAT,            // códec (p.ej. alaw/ulaw/slin16)
      transport: EXTERNAL_TRANSPORT,      // udp/tcp/websocket
      encapsulation: EXTERNAL_ENCAPSULATION, // rtp/audiosocket/none
      connection_type: 'client',
      direction: EXTERNAL_DIRECTION,      // both/in/out
    };

    const extCh = await client.channels.externalMedia(params);
    CALLS.get(sipId).extChannelId = extCh.id;
    EXT_TO_SIP.set(extCh.id, sipId);
    log.info(`ExternalMedia creado: ${extCh.id} -> ${EXTERNAL_HOST} (${EXTERNAL_FORMAT}/${EXTERNAL_TRANSPORT}/${EXTERNAL_ENCAPSULATION})`);

  } catch (e) {
    log.error(`Error preparando llamada para SIP ${sipId}: ${e.message}`);
    try { await ch.hangup(); } catch {}
    await cleanupCall(sipId);
  }
}

async function onStasisEnd(event, channel) {
  const chId   = safeGet(event, ['channel', 'id']);
  const chName = safeGet(event, ['channel', 'name']);
  log.info(`StasisEnd: channel_id=${chId} name=${chName}`);

  // Si terminó un ExternalMedia, solo desmapear y salir
  if (EXT_TO_SIP.has(chId)) {
    const sipId = EXT_TO_SIP.get(chId);
    EXT_TO_SIP.delete(chId);
    log.info(`External ${chId} terminó (SIP asociado: ${sipId})`);
    return;
  }

  // Si terminó un SIP, limpiar todo
  await cleanupCall(chId);
}

// --- Limpieza de recursos por llamada ---
async function cleanupCall(sipId) {
  const state = CALLS.get(sipId);
  if (!state) return;

  // detener límite
  if (state.limitTimer) {
    clearInterval(state.limitTimer);
    state.limitTimer = null;
  }

  // colgar external si existe
  if (state.extChannelId) {
    try {
      await client.channels.hangup({ channelId: state.extChannelId });
      log.info(`External ${state.extChannelId} colgado`);
    } catch {}
    EXT_TO_SIP.delete(state.extChannelId);
  }

  // destruir bridge si existe
  if (state.bridgeId) {
    try {
      const br = await client.bridges.get({ bridgeId: state.bridgeId });
      await br.destroy();
      log.info(`Bridge ${state.bridgeId} destruido`);
    } catch {}
  }

  // colgar SIP (por si sigue vivo)
  try {
    await client.channels.hangup({ channelId: sipId });
  } catch {}

  CALLS.delete(sipId);
}

// --- Señales ---
async function shutdown(sig) {
  log.info(`Señal ${sig} recibida. Cerrando…`);
  try {
    // limpiar todas las llamadas activas
    const ids = Array.from(CALLS.keys());
    for (const id of ids) {
      await cleanupCall(id);
    }
  } finally {
    try { client && client.close && client.close(); } catch {}
    process.exit(0);
  }
}

process.on('SIGINT',  () => shutdown('SIGINT'));
process.on('SIGTERM', () => shutdown('SIGTERM'));

// --- Main ---
(async () => {
  if (!ARI_URL) {
    log.error('Falta ARI_URL en variables de entorno.');
    process.exit(1);
  }
  if (!EXTERNAL_HOST) {
    log.error('Falta EXTERNAL_HOST (host:port del gateway).');
    process.exit(1);
  }

  try {
    log.info(`Conectando a ARI: ${ARI_URL} (app=${ARI_APP})`);
    client = await ari.connect(ARI_URL, ARI_USER, ARI_PASS);

    // Suscripción a eventos
    client.on('StasisStart', onStasisStart);
    client.on('StasisEnd',   onStasisEnd);
    client.on('error', (err) => log.error(`ARI error: ${err.message}`));
    client.on('close', () => log.info('Conexión ARI cerrada'));

    // Iniciar la app
    await client.start(ARI_APP);
    log.info('Escuchando eventos…');
  } catch (e) {
    log.error(`No se pudo conectar o iniciar ARI: ${e.message}`);
    process.exit(1);
  }
})();

// --- Helpers ---
function delay(ms){ return new Promise(r => setTimeout(r, ms)); }
